rocAL/source/readers/image/numpy_data_reader.cpp

```diff
@@ -0,0 +1,578 @@
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include <math.h>
+
+#include <algorithm>
+#include <cassert>
+#include <cstring>
+#include <numeric>
+#include <random>
+
+#include "pipeline/commons.h"
+#include "pipeline/filesystem.h"
+#include "readers/image/numpy_data_reader.h"
+
+NumpyDataReader::NumpyDataReader() {
+    _src_dir = nullptr;
+    _sub_dir = nullptr;
+    _entity = nullptr;
+    _curr_file_idx = 0;
+    _current_file_size = 0;
+    _current_fPtr = nullptr;
+    _loop = false;
+    _shuffle = false;
+    _file_count_all_shards = 0;
+}
+
+unsigned NumpyDataReader::count_items() {
+    int size = get_max_size_of_shard(_batch_size, _loop);
+    int ret = (size - _read_counter);
+    if (_sharding_info.last_batch_policy == RocalBatchPolicy::DROP && _last_batch_padded_size != 0)
+        ret -= _batch_size;
+    return ((ret < 0) ? 0 : ret);
+}
+
+Reader::Status NumpyDataReader::initialize(ReaderConfig desc) {
+    auto ret = Reader::Status::OK;
+    _folder_path = desc.path();
+    _file_list_path = desc.file_list_path();
+    _shard_id = desc.get_shard_id();
+    _shard_count = desc.get_shard_count();
+    _batch_size = desc.get_batch_size();
+    _shuffle = desc.shuffle();
+    _loop = desc.loop();
+    _meta_data_reader = desc.meta_data_reader();
+    _sharding_info = desc.get_sharding_info();
+    _pad_last_batch_repeated = _sharding_info.pad_last_batch_repeated;
+    _stick_to_shard = _sharding_info.stick_to_shard;
+    _shard_size = _sharding_info.shard_size;
+    _files = desc.get_files();
+    _seed = desc.seed();
+    ret = subfolder_reading();
+    _file_headers.resize(_file_names.size());
+    // shuffle dataset if set
+    if (ret == Reader::Status::OK && _shuffle) {
+        std::mt19937 rng(_seed);
+        std::shuffle(_file_names.begin() + _shard_start_idx_vector[_shard_id],
+                     _file_names.begin() + _shard_end_idx_vector[_shard_id], rng);
+    }
+    return ret;
+}
+
+void NumpyDataReader::incremenet_read_ptr() {
+    _read_counter++;
+    increment_curr_file_idx(_file_names.size());
+}
+
+size_t NumpyDataReader::open() {
+    auto file_path = _file_names[_curr_file_idx];       // Get current file name
+    _curr_file_header = _file_headers[_curr_file_idx];  // Get current file header
+    incremenet_read_ptr();
+    _last_file_path = _last_id = file_path;
+    auto last_slash_idx = _last_id.find_last_of("\\/");
```

Comment:
Can we move these to std::string id() function? Since _last_id is not used anywhere here.